packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
+            assert(existing.line == null);
+            existing.line = lines[i];
+            assert(existing.column == null);
+            existing.column = columns[i];
+            assert(existing.name == null);
+            existing.name = names[i];
+            _countUnkownLocations--;
+          } else {
+            // Existing entry already has a location. Ensure it is consistent.
+            // Data could become inconsistent if we had a bug and comingled data
+            // from before and after a hot restart.
+            assert(existing.path == path);
+            assert(existing.line == lines[i]);
+            assert(existing.column == columns[i]);
+            assert(existing.name == names[i]);
+          }
+        } else {
+          final location = Location(
+            id: id,
+            path: path,
             line: lines[i],
             column: columns[i],
             name: names[i],
           );
+
           _locationMap[ids[i]] = location;
-          _locations.add(location);
         }
       }
-    } else if (json.containsKey('newLocations')) {
-      // Fall back to the older 'newLocations' field.
-      final fileLocationsMap =
-          (json['newLocations'] as Map).cast<String, dynamic>();
+    }
 
-      for (final file in fileLocationsMap.keys) {
-        final List<int> entries = (fileLocationsMap[file] as List).cast<int>();
+    _locationsResolved.value = _countUnkownLocations == 0;
+  }
+}
 
-        final shortName = path.posix.split(file).last;
+class RebuildCountModel {
+  RebuildCountModel();
 
-        final len = entries.length ~/ 3;
-        for (var i = 0; i < (len * 3); i += 3) {
-          final id = entries[i];
-          final line = entries[i + 1];
-          final location = RebuildLocation(
-            id: id,
-            path: file,
-            line: line,
-            column: entries[i + 2],
-            name: '$shortName:$line',
-          );
-          _locationMap[id] = location;
-          _locations.add(location);
-        }
+  RebuildCountModel.fromJson(Map<String, dynamic> json) {
+    if (json.isEmpty) return;
+    locationMap.processLocationMap(json['locations']);
+    final frames =
+        (json['frames'] as List<dynamic>).cast<Map<String, dynamic>>();
+    frames.forEach(processRebuildsForFrame);
+  }
+
+  // Maximum number of historic frames to keep rebuild counts to ensure memory
+  // usage from rebuild counts is not excessive.
+  static const int rebuildFrameCacheSize = 10000;
+
+  /// Source of truth for all resolution fo location ids to [Location] objects.
+  final LocationMap locationMap = LocationMap();
+
+  /// Map from frame id to list of rebuilds for the frame.
+  final _rebuildsForFrame = <int, List<RebuildLocation>>{};
+
+  final _stats = _RebuildCountStats();
+  ValueListenable<List<RebuildLocation>> get locationStats => _stats.locations;
+
+  List<RebuildLocation>? get rebuildsForLastFrame =>
+      _rebuildsForFrame.values.lastOrNull;
```

Comment:
is .values guaranteed to be in the order that items were added to the map? is this the order we want?

Reply:
yep. because of the kind of map used.